File: ui/app/types/book.ts
export interface Book {
  id: string;
  title: string;
  imageUrl: string;
  isRead: boolean;
  author?: string;
  description?: string;
  genre?: string;
  publishedYear?: number;
  productUrl?: string;
}

export interface ProductAttribute {
  name: string;
  value: string;
}

export interface Product {
  sku: string;
  body: string;
  title: string;
  image_url: string;
  product_url: string;
  attributes: ProductAttribute[];
}

export interface Card {
  type: string;
  id: string;
  product: Product;
}

export interface LingerEventProperties {
  organization_id: string;
  visitor_id: string;
  session_id: string;
  payload: Record<string, {
    enter_count: number;
    id: string;
    time: number;
    type: string;
  }>;
}

export interface AbstractInterestProperties {
  organization_id: string;
  visitor_id: string;
  session_id: string;
  id: string;
  weight: number;
}

export interface Event {
  event: string;
  properties: LingerEventProperties | AbstractInterestProperties;
}

export interface RecommendationRequest {
  sessionId: string;
  page?: number;
  batchCount?: number;
  events?: Event[];
  searchPrompt?: string;
}

export interface RecommendationResponse {
  cards: Card[];
}

File: apps/ios/DeweyApp/DeweyApp/Models/Book.swift
import Foundation

enum ReadingStatus: String, Codable {
    case wantToRead = "want_to_read"
    case reading
    case read
}

struct Book: Codable, Identifiable, Hashable {
    let id: String
    var title: String
    var imageURL: URL?
    var isRead: Bool
    var status: ReadingStatus?
    var author: String?
    var description: String?
    var genre: String?
    var publishedYear: Int?
    var productURL: URL?

    init(
        id: String,
        title: String,
        imageURL: URL?,
        isRead: Bool,
        status: ReadingStatus?,
        author: String?,
        description: String?,
        genre: String?,
        publishedYear: Int?,
        productURL: URL?
    ) {
        self.id = id
        self.title = title
        self.imageURL = imageURL
        self.isRead = isRead
        self.status = status
        self.author = author
        self.description = description
        self.genre = genre
        self.publishedYear = publishedYear
        self.productURL = productURL
    }
}
